Feature/first-in-voxel filter

pravinshinde825 opened the pull request:
- Added `first-in-voxel filter`.
      - This filter will do a pointcloud downsizing based on voxel-grid. 
      - One of the limitation to the `voxelcenternearestneibour` filter is it cannot run in streamable mode. To overcome the limitation, I wrote first-in-voxel filter. This filter will only accept the first point entering in the voxel and mark the voxel as populated, After that any point entering in the same voxel will be ignored/skipped.

nicolas-chaulet commented:
> I guess I'm not convinced the center approach couldn't be made streamable (see comments in #2629), but if this works for you then I have no objections. I can take a closer look at the code today.

The voxel centre is streamable and could be an option to this filter I guess. The one that is not streamable is the voxel centroid :)

pravinshinde825 commented on filters/FirstInVoxelFilter.cpp:
```diff
@@ -73,24 +73,53 @@ PointViewSet FirstInVoxelFilter::run(PointViewPtr view)
+    size_t const hx(std::hash<int>{}(std::floor(gx - m_Bounds.minx)));
+    size_t const hy(std::hash<int>{}(std::floor(gy - m_Bounds.miny)));
+    size_t const hz(std::hash<int>{}(std::floor(gz - m_Bounds.minz)));
+    size_t key = std::hash<size_t>{}(HASH_COMBINE((HASH_COMBINE(hx, hy)), hz));
```
Updated to use `std::tuple`.

pravinshinde825 commented on filters/FirstInVoxelFilter.cpp:
```diff
@@ -0,0 +1,132 @@
+namespace pdal
+{
+
+static StaticPluginInfo const s_info{"filters.first-in-voxel",
```
Done!!

pravinshinde825 commented on filters/FirstInVoxelFilter.cpp:
```diff
@@ -0,0 +1,132 @@
+
+void FirstInVoxelFilter::addArgs(ProgramArgs& args)
+{
+    args.add("cell", "Cell size", m_cell).setPositional();
```
done!!

pravinshinde825 commented on filters/FirstInVoxelFilter.cpp:
```diff
@@ -73,24 +73,53 @@ PointViewSet FirstInVoxelFilter::run(PointViewPtr view)
+		* Save global coordinates of first incoming point. 
+		* This will act as a Pivot for calculation of local coordinates of the voxels.
+		*/
+		m_Bounds.grow(gx, gy, gz);
```
Changed to array.

pravinshinde825 commented on filters/FirstInVoxelFilter.cpp:
```diff
@@ -0,0 +1,102 @@
+
+    m_Bounds.grow(gx, gy, gz);
+
+    size_t const h1(std::hash<int>{}(static_cast<uint32_t>(POSITIVE_VALUE((gx - m_Bounds.minx)))));
```
Not needed.

pravinshinde825 commented:
@chambbj , Sorry for the delayed response.
I addressed comments and updated code accordingly. Below is a snapshot for output according to latest changes.
![Capture](https://user-images.githubusercontent.com/38933204/64229429-5d62fd80-cf07-11e9-8920-97cffaf58f94.JPG)

nicolas-chaulet commented on filters/FirstInVoxelFilter.hpp:
```diff
@@ -0,0 +1,68 @@
+    virtual PointViewSet run(PointViewPtr view) override;
+    bool processOne(PointRef& point) override;
+    bool voxelize(const PointRef point);
+    std::set<std::tuple<int,int,int>> m_populatedVoxels;
```
@pravinshinde825 I think we would get some nice speed boosts by making that one an `unordered_set`, did you look into that?

pravinshinde825 replied:
@nicolas-chaulet , Yes I checked with unordered_set as well. 
- In performance point of view `unordered_set` is a bit faster than `std::set` but not much. 
- In memory point of view I noticed approx 60% more memory usage in `unordered_set` than `set`.
- Also we may need to inject a hash funtion for `std::tuple<int,int,int>`.
So, I think `set` is fine.

chambbj commented on filters/FirstInVoxelFilter.cpp:
```diff
@@ -0,0 +1,130 @@
+{
+
+static StaticPluginInfo const s_info{"filters.firstInVoxel",
+                                     "First Entry Voxel Filter", ""};
```
Third (currently empty field) should contain link to the docs, which will be http://pdal.io/stages/filters.first-in-voxel.html as things stand currently, though I think the doc page should be updated to match the name of the filter (having dropped the hyphens now).

pravinshinde825 replied:
Added link to the doc.

chambbj commented on doc/stages/filters.first-in-voxel.rst:
```diff
@@ -0,0 +1,37 @@
+-------------------------------------------------------------------------------
+
+cell
+  Cell size in the ``X``, ``Y``, and ``Z`` dimension.
```
Please indicate the default value for the cell size by adding `[Default: 0.001]` to the end of the description.

pravinshinde825 replied:
Added link to the doc.

chambbj commented on test/unit/filters/FirstInVoxelFilterTest.cpp:
```diff
@@ -0,0 +1,165 @@
+#include <pdal/pdal_test_main.hpp>
+
+#include <pdal/StageFactory.hpp>
+#include<pdal/Reader.hpp>
```
These three lines should have spaces after `#include` for consistency.

pravinshinde825 replied:
Added spaces after `#include`

chambbj commented:
@pravinshinde825 Things are looking pretty good I think. I get believable results now. Just a few minor edits requested and then you should be good to go.

abellgithub commented:
I just wonder what, if any, advantage this has over the existing filter.  Could this logic, along with a check for distance, be added to VoxelCenterNearestNeighbor?  Is there some functional advantage to choosing the first point seen, rather than the one nearest to the center of the voxel?

nicolas-chaulet commented:
@abellgithub I can't thing of a way to get VoxelCenterNearestNeighbor to work in streaming mode. You need to go through the whole file before you can decide which point to keep. The idea here was to have a voxel downsizing that works in streaming mode. Same filter could be used to pick the centre just like PCL does it (as opposed to the first point)

chambbj commented:
I think it's as simple as, when you encounter an already initialized voxel, check to see if the distance for the current point is smaller than that of the one that first initialized the voxel, and if it is, update the dimensions/swap the point.

nicolas-chaulet commented:
If you are streaming, how can you swap? The first point has already been given to the next stage no?

chambbj commented:
This is true. Sorry.

abellgithub commented:
Sorry.  For some reason I was thinking of this filter as a writer -- that the voxel set was the output.

abellgithub commented on filters/FirstInVoxelFilter.cpp:
```diff
@@ -0,0 +1,131 @@
+    int gy = point.getFieldAs<double>(Dimension::Id::Y) / m_cell;
+    int gz = point.getFieldAs<double>(Dimension::Id::Z) / m_cell;
+
+    static bool initialized = false;
```
This needs to be moved to be a member variable or multiple instances of this filter will be confused.

pravinshinde825 replied:
Yeah, make sense.
Changed to member variable

abellgithub commented on filters/FirstInVoxelFilter.cpp:
```diff
@@ -0,0 +1,131 @@
+     * Calculate the local voxel coordinates for incoming point, Using the Pivot
+     * voxel.
+     */
+    auto t=std::make_tuple(gx - m_pivotVoxel[0], gy - m_pivotVoxel[1],
```
Spacing

pravinshinde825 replied:
Done!!

pravinshinde825 commented:
Thanks @nicolas-chaulet for the accurate description.
@chambbj , @abellgithub , I addressed all the comments.
